### PR TITLE
issue #930 - ensure status is never null from delete AND return 200 OK on success

### PR DESCRIPTION
**Pull request by lmsurpre**
While investigating the issue, I found that `delete` and `conditionalDelete` were handled subtely different.
Now the response status is set in FHIRRestHelper.doDelete() and these two variants share a common buildResponse helper.

Additionally, we were inconsistent in the response status being returned.
Previously we were returning 204 (No Content) for some deletes and 200 (OK) status for others.
Now we will always return a 200 (OK) for a successful delete and the response will always contain an OperationOutcome:
* For conditional deletes with no matches, we return a warning
* For the deletion of a previously deleted resource, we return a warning
* For the deletion of a non-existent resource, we return a warning
* For any other successful deletion (conditional or not), we return an informational message with the number of resources deleted (and their logical ids)

Signed-off-by: Lee Surprenant <lmsurpre@us.ibm.com>

### CODE REVIEW

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 1]
```diff
@@ -567,75 +567,77 @@ private ResourceDAO getResourceDao() {
...
-                    resourceBuilder = existingResource.toBuilder();
-
-                    SingleResourceResult<T> result = new SingleResourceResult.Builder<T>()
-                            .success(true)
-                            .resource(existingResource)
-                            .build();
+            if (existingResourceDTO == null) {
+                throw new FHIRPersistenceResourceNotFoundException("resource does not exist: " + 
+                        resourceType.getSimpleName() + ":" + logicalId);
+            }
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:573]: the changes here are mostly just for style:  by moving the exception cases to the top, i reduced the main code block by two levels of indentation

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 2]
```diff
@@ -567,75 +567,77 @@ private ResourceDAO getResourceDao() {
...
-                    existingResource = this.convertResourceDTO(existingResourceDTO, resourceType, null);
-                    resourceBuilder = existingResource.toBuilder();
-
-                    SingleResourceResult<T> result = new SingleResourceResult.Builder<T>()
-                            .success(true)
-                            .resource(existingResource)
-                            .build();
+            if (existingResourceDTO == null) {
+                throw new FHIRPersistenceResourceNotFoundException("resource does not exist: " + 
+                        resourceType.getSimpleName() + ":" + logicalId);
```
Comment (albertwang-ibm) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:572]: 2 ":" seems a little bit strange, can we change to 
resourceType.getSimpleName() + ":" + logicalId" + "does not exist." 

or use the same style as the other warning info of the following codes.

**Comment by albertwang-ibm** [comment]
this should be able to pass touchstone tests, touch stone only verifies: if 200 then should have body, if 204, then no body.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 2 above)
Comment (prb112) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:572]: Adding to Albert's comments... we should bracket the values in the message.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 3]
```diff
@@ -893,8 +902,7 @@ private ResourceDAO getResourceDao() {
      * @throws IOException
      */
     protected List<Resource> buildSortedFhirResources(FHIRPersistenceContext context, Class<? extends Resource> resourceType, List<Long> sortedIdList,
-                                                      List<String> elements)
-                            throws FHIRException, FHIRPersistenceException, IOException {
+            List<String> elements) throws FHIRException, FHIRPersistenceException, IOException {
         final String METHOD_NAME = "buildFhirResource";
```
Comment (prb112) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:906]: wrong method name

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 2 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:572]: My PR didn't actually change the error message from before, but I agree with these comments and will address now.

Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:572]: I think I'll just change it to `"resource does not exist: " + resourceType.getSimpleName() + "/" + logicalId)`
I usually like to bracket the values in `'` but in this case the values can't contain spaces

Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:572]:
```suggestion
                        resourceType.getSimpleName() + "/" + logicalId);
```

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 3 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:906]:
```suggestion
        final String METHOD_NAME = "buildSortedFhirResources";
```

**Pull request merged**